In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/european-airlines-catastrophes/metadata.txt
/kaggle/input/european-airlines-catastrophes/sample.csv
/kaggle/input/european-airlines-catastrophes/train.csv
/kaggle/input/european-airlines-catastrophes/test.csv


In [2]:
train_df = pd.read_csv("../input/european-airlines-catastrophes/train.csv")
train_df

Id  inspectionScore  maneuverabilityScore  maintananceConcerns  \
0        0        16.301370             59.161349                   17   
1        1        18.310502             62.534184                    8   
2        2        25.662100             70.738377                   10   
3        3        37.488584             66.636281                    7   
4        4        58.401826             71.741112                    4   
...    ...              ...                   ...                  ...   
8995  8995        82.283105             66.134913                    0   
8996  8996        37.579909             60.802188                   10   
8997  8997        24.611872             82.041933                    4   
8998  8998        38.036530             72.151322                    4   
8999  8999        39.452055             59.298086                    3   

      weatherSeverityScore  temperature  safetyCheckFailures  galeStrength  \
0                 0.129040        79.66                    3      0.298803   
1                 0.003186        78.00                    1      0.288345   
2                 0.064877        79.17                    2      0.207206   
3                 0.003557        81.56                    2      0.295558   
4                 0.356823        79.31                    2      0.362273   
...                    ...          ...                  ...           ...   
8995              0.009234        77.18                    1      0.335587   
8996              0.849289        76.73                    3      0.370567   
8997              0.001419        80.14                    1      0.319359   
8998              0.007158        78.30                    2      0.314310   
8999              0.002341        79.24                    2      0.316834   

      lastInspected  incidentCode  incidentID  maxCruisingElevation  \
0                16             3       10888          26116.134006   
1                16             7         587          36059.221584   
2                14             4        4574          35594.722894   
3                14             6       12044          14871.664920   
4                11             2        8948          26391.631688   
...             ...           ...         ...                   ...   
8995              7             6        5516          34982.992727   
8996             10             1        4599          23171.545088   
8997             22             7        4025          13945.572866   
8998             14             6        8945          29519.010231   
8999             13             7         195          26226.196218   

     scaleOfIncident  
0       catastrophic  
1       catastrophic  
2       catastrophic  
3             medium  
4                low  
...              ...  
8995          medium  
8996    catastrophic  
8997             low  
8998          medium  
8999          medium  

[9000 rows x 13 columns]

In [3]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Id                    9000 non-null   int64  
 1   inspectionScore       9000 non-null   float64
 2   maneuverabilityScore  9000 non-null   float64
 3   maintananceConcerns   9000 non-null   int64  
 4   weatherSeverityScore  9000 non-null   float64
 5   temperature           9000 non-null   float64
 6   safetyCheckFailures   9000 non-null   int64  
 7   galeStrength          9000 non-null   float64
 8   lastInspected         9000 non-null   int64  
 9   incidentCode          9000 non-null   int64  
 10  incidentID            9000 non-null   int64  
 11  maxCruisingElevation  9000 non-null   float64
 12  scaleOfIncident       9000 non-null   object 
dtypes: float64(6), int64(6), object(1)
memory usage: 914.2+ KB


In [4]:
train_df.isnull().sum()

Id                      0
inspectionScore         0
maneuverabilityScore    0
maintananceConcerns     0
weatherSeverityScore    0
temperature             0
safetyCheckFailures     0
galeStrength            0
lastInspected           0
incidentCode            0
incidentID              0
maxCruisingElevation    0
scaleOfIncident         0
dtype: int64

In [5]:
train_df['scaleOfIncident'].value_counts()

catastrophic    2754
medium          2472
low             2267
high            1507
Name: scaleOfIncident, dtype: int64

In [6]:
train_df.groupby('scaleOfIncident').mean()

Id  inspectionScore  maneuverabilityScore  \
scaleOfIncident                                                       
catastrophic     4586.654321        33.308297             66.442562   
high             4500.994028        47.229944             60.181233   
low              4389.520512        47.760221             64.381987   
medium           4502.351537        42.650674             67.455373   

                 maintananceConcerns  weatherSeverityScore  temperature  \
scaleOfIncident                                                           
catastrophic                6.321714              0.304583    80.022059   
high                        6.568679              0.244077    80.036523   
low                         6.830172              0.313243    79.907442   
medium                      6.635922              0.157594    79.911351   

                 safetyCheckFailures  galeStrength  lastInspected  \
scaleOfIncident                                                     
catastrophic                1.957153      0.367512      12.769789   
high                        2.180491      0.411693      13.105508   
low                         1.986326      0.386329      13.189237   
medium                      1.991909      0.373539      12.834547   

                 incidentCode   incidentID  maxCruisingElevation  
scaleOfIncident                                                   
catastrophic         3.469862  6263.306100          31892.785785  
high                 4.518912  6161.923689          31394.149546  
low                  3.265990  6317.971328          32277.043286  
medium               4.282767  6230.190534          32302.577857

In [7]:
train_df.columns.values

array(['Id', 'inspectionScore', 'maneuverabilityScore',
       'maintananceConcerns', 'weatherSeverityScore', 'temperature',
       'safetyCheckFailures', 'galeStrength', 'lastInspected',
       'incidentCode', 'incidentID', 'maxCruisingElevation',
       'scaleOfIncident'], dtype=object)

In [8]:
train_df = train_df.drop(columns = 'Id')

In [9]:
train_df

inspectionScore  maneuverabilityScore  maintananceConcerns  \
0           16.301370             59.161349                   17   
1           18.310502             62.534184                    8   
2           25.662100             70.738377                   10   
3           37.488584             66.636281                    7   
4           58.401826             71.741112                    4   
...               ...                   ...                  ...   
8995        82.283105             66.134913                    0   
8996        37.579909             60.802188                   10   
8997        24.611872             82.041933                    4   
8998        38.036530             72.151322                    4   
8999        39.452055             59.298086                    3   

      weatherSeverityScore  temperature  safetyCheckFailures  galeStrength  \
0                 0.129040        79.66                    3      0.298803   
1                 0.003186        78.00                    1      0.288345   
2                 0.064877        79.17                    2      0.207206   
3                 0.003557        81.56                    2      0.295558   
4                 0.356823        79.31                    2      0.362273   
...                    ...          ...                  ...           ...   
8995              0.009234        77.18                    1      0.335587   
8996              0.849289        76.73                    3      0.370567   
8997              0.001419        80.14                    1      0.319359   
8998              0.007158        78.30                    2      0.314310   
8999              0.002341        79.24                    2      0.316834   

      lastInspected  incidentCode  incidentID  maxCruisingElevation  \
0                16             3       10888          26116.134006   
1                16             7         587          36059.221584   
2                14             4        4574          35594.722894   
3                14             6       12044          14871.664920   
4                11             2        8948          26391.631688   
...             ...           ...         ...                   ...   
8995              7             6        5516          34982.992727   
8996             10             1        4599          23171.545088   
8997             22             7        4025          13945.572866   
8998             14             6        8945          29519.010231   
8999             13             7         195          26226.196218   

     scaleOfIncident  
0       catastrophic  
1       catastrophic  
2       catastrophic  
3             medium  
4                low  
...              ...  
8995          medium  
8996    catastrophic  
8997             low  
8998          medium  
8999          medium  

[9000 rows x 12 columns]

In [10]:
X = train_df.drop(columns='scaleOfIncident')
y = train_df['scaleOfIncident']

In [11]:
X

inspectionScore  maneuverabilityScore  maintananceConcerns  \
0           16.301370             59.161349                   17   
1           18.310502             62.534184                    8   
2           25.662100             70.738377                   10   
3           37.488584             66.636281                    7   
4           58.401826             71.741112                    4   
...               ...                   ...                  ...   
8995        82.283105             66.134913                    0   
8996        37.579909             60.802188                   10   
8997        24.611872             82.041933                    4   
8998        38.036530             72.151322                    4   
8999        39.452055             59.298086                    3   

      weatherSeverityScore  temperature  safetyCheckFailures  galeStrength  \
0                 0.129040        79.66                    3      0.298803   
1                 0.003186        78.00                    1      0.288345   
2                 0.064877        79.17                    2      0.207206   
3                 0.003557        81.56                    2      0.295558   
4                 0.356823        79.31                    2      0.362273   
...                    ...          ...                  ...           ...   
8995              0.009234        77.18                    1      0.335587   
8996              0.849289        76.73                    3      0.370567   
8997              0.001419        80.14                    1      0.319359   
8998              0.007158        78.30                    2      0.314310   
8999              0.002341        79.24                    2      0.316834   

      lastInspected  incidentCode  incidentID  maxCruisingElevation  
0                16             3       10888          26116.134006  
1                16             7         587          36059.221584  
2                14             4        4574          35594.722894  
3                14             6       12044          14871.664920  
4                11             2        8948          26391.631688  
...             ...           ...         ...                   ...  
8995              7             6        5516          34982.992727  
8996             10             1        4599          23171.545088  
8997             22             7        4025          13945.572866  
8998             14             6        8945          29519.010231  
8999             13             7         195          26226.196218  

[9000 rows x 11 columns]

In [12]:
y

0       catastrophic
1       catastrophic
2       catastrophic
3             medium
4                low
            ...     
8995          medium
8996    catastrophic
8997             low
8998          medium
8999          medium
Name: scaleOfIncident, Length: 9000, dtype: object

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [14]:
X_train.shape

(6300, 11)

In [15]:
X_test.shape

(2700, 11)

In [16]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

# **Hyperparameter optimization**

In [23]:
n_estimators = [int(n) for n in np.linspace(start = 1, stop = 500, num = 10)]
criterion = ['gini', 'entropy']
max_depth = [10,20,5,3]
min_sample_leaf = [5,7,10,15,20]
max_features = ['auto','sqrt','log2']

In [24]:
 
hyper_parameter = {
    'n_estimators' : n_estimators,
    'criterion' : criterion,
    'max_depth': max_depth,
    'min_samples_leaf': min_sample_leaf,
    'max_features':max_features
                   
}

# **Using RandomForestClassifier**

In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

model = RandomForestClassifier()

In [26]:
RS = RandomizedSearchCV(estimator = model, param_distributions = hyper_parameter,n_iter = 10, cv = 5, )

In [27]:
RS.fit(X_train,y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [10, 20, 5, 3],
                                        'max_features': ['auto', 'sqrt',
                                                         'log2'],
                                        'min_samples_leaf': [5, 7, 10, 15, 20],
                                        'n_estimators': [1, 56, 111, 167, 222,
                                                         278, 333, 389, 444,
                                                         500]})

In [29]:
y_predictions = RS.predict(X_test)

In [30]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_predictions)
print(confusion_matrix)

[[738  28  47  32]
 [  5 364  40  10]
 [ 15  17 630  27]
 [ 36   6  63 642]]


In [31]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predictions))

              precision    recall  f1-score   support

catastrophic       0.93      0.87      0.90       845
        high       0.88      0.87      0.87       419
         low       0.81      0.91      0.86       689
      medium       0.90      0.86      0.88       747

    accuracy                           0.88      2700
   macro avg       0.88      0.88      0.88      2700
weighted avg       0.88      0.88      0.88      2700



#   **RandomForestClassifier gives an accuracy of 88%**

# **Using XGBooster along with hyperparameter tuning**

In [17]:
import matplotlib.pyplot as plt

from xgboost import XGBClassifier
from xgboost import XGBRFClassifier
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

In [18]:
def hyperParameterTuning(X_train, y_train):
    param_tuning = {
        'learning_rate': [0.01, 0.1],
        'max_depth': [3, 5, 7, 10],
        'min_child_weight': [1, 3, 5],
        'subsample': [0.5, 0.7],
        'colsample_bytree': [0.5, 0.7],
        'n_estimators' : [100, 200, 500],
        'objective': ['reg:squarederror']
    }
    XGB = XGBClassifier()
    
    gridsearch = GridSearchCV(estimator = XGB,
                           param_grid = param_tuning,                        
                           cv = 5,
                           n_jobs = -1,
                           verbose = 1)

    gridsearch.fit(X_train,y_train)

    return gridsearch.best_params

In [19]:
XGB = XGBClassifier(
        objective = 'reg:squarederror',
        colsample_bytree = 0.5,
        learning_rate = 0.05,
        max_depth = 6,
        min_child_weight = 1,
        n_estimators = 1000,
        subsample = 0.7)

%time XGB.fit(X_train, y_train, early_stopping_rounds=5, eval_set=[(X_test, y_test)], verbose=False)

/opt/conda/lib/python3.7/site-packages/xgboost/sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[22:44:01] WARNING: ../src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
CPU times: user 32.4 s, sys: 276 ms, total: 32.7 s
Wall time: 9.12 s


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.05, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=1000, n_jobs=4, num_parallel_tree=1,
              objective='multi:softprob', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=None, subsample=0.7,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [20]:
y_preds = XGB.predict(X_test)

In [21]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_preds)
print(confusion_matrix)

[[769  22  38  16]
 [ 15 387  10   7]
 [  9   8 658  14]
 [ 13  15  41 678]]


In [22]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_preds))

              precision    recall  f1-score   support

catastrophic       0.95      0.91      0.93       845
        high       0.90      0.92      0.91       419
         low       0.88      0.96      0.92       689
      medium       0.95      0.91      0.93       747

    accuracy                           0.92      2700
   macro avg       0.92      0.92      0.92      2700
weighted avg       0.92      0.92      0.92      2700



# **XGBooster gives an accuracy of 92% and proved to be more effective than RandomForestClassifier**